lte/cloud/go/services/subscriberdb/storage/sub_proto.go

```diff
@@ -0,0 +1,335 @@
+					From(subProtoTmpTableName).
+					Where(squirrel.Eq{subProtoNidCol: network}),
+			).
+			OnConflict(
+				[]sqorc.UpsertValue{{
+					Column: subProtoProtoCol,
+					Value:  squirrel.Expr(fmt.Sprintf("%s.%s", upsertColumnPrefix, subProtoProtoCol)),
+				}},
+				subProtoNidCol, subProtoSidCol,
+			).
+			RunWith(tx).
+			Exec()
+		if err != nil {
+			return nil, errors.Wrapf(err, "populate sub proto store table")
+		}
+
+		_, err = l.builder.
+			Delete(subProtoTmpTableName).
+			Where(squirrel.Eq{subProtoNidCol: network}).
+			RunWith(tx).
+			Exec()
+		if err != nil {
+			return nil, errors.Wrapf(err, "clean up tmp sub proto store table")
+		}
+		return nil, nil
+	}
+	_, err := sqorc.ExecInTx(l.db, nil, nil, txFn)
+	return err
+}
+
+func (l *SubProtoStore) ClearTmpTable() error {
+	txFn := func(tx *sql.Tx) (interface{}, error) {
+		_, err := l.builder.
+			Delete(subProtoTmpTableName).
+			RunWith(tx).
+			Exec()
+		return nil, errors.Wrapf(err, "clear sub protos tmp table")
+	}
+	_, err := sqorc.ExecInTx(l.db, nil, nil, txFn)
+	return err
+}
+
+// DeleteSubProtos deletes the cached protos for a list of networks.
+func (l *SubProtoStore) DeleteSubProtos(networks []string) error {
+	txFn := func(tx *sql.Tx) (interface{}, error) {
+		_, err := l.builder.
+			Delete(subProtoTableName).
+			Where(squirrel.Eq{subProtoNidCol: networks}).
+			RunWith(tx).
+			Exec()
+		return nil, errors.Wrapf(err, "delete sub protos for networks %+v", networks)
+	}
+	_, err := sqorc.ExecInTx(l.db, nil, nil, txFn)
+	return err
+}
+
+// GetPage gets a page of subscriber protos based on the page token and size, and also returns
+// the token for the next page.
+func (l *SubProtoStore) GetPage(network string, token string, pageSize uint64) ([]*lte_protos.SubscriberData, string, error) {
+	lastIncludedSid := ""
+	if token != "" {
+		decoded, err := decodePageToken(token)
+		if err != nil {
+			return nil, "", err
+		}
+		lastIncludedSid = decoded.LastIncludedEntity
+	}
+	txFn := func(tx *sql.Tx) (interface{}, error) {
+		rows, err := l.builder.
+			Select(subProtoProtoCol).
+			From(subProtoTableName).
+			Where(squirrel.And{
+				squirrel.Eq{subProtoNidCol: network},
+				squirrel.Gt{subProtoSidCol: lastIncludedSid},
+			}).
+			OrderBy(subProtoSidCol).
+			Limit(pageSize).
+			RunWith(tx).
+			Query()
+		if err != nil {
+			return nil, errors.Wrapf(err, "get page for network %+v with token %+v", network, token)
+		}
+		return parseSubProtoRows(rows)
+	}
+	ret, err := sqorc.ExecInTx(l.db, nil, nil, txFn)
+	if err != nil {
+		return nil, "", err
+	}
+	subProtos := ret.([]*lte_protos.SubscriberData)
+	nextPageToken, err := getNextPageToken(subProtos)
+	if err != nil {
+		return nil, "", errors.Wrapf(err, "get next page token")
+	}
+	return subProtos, nextPageToken, nil
+}
+
+// GetByIDs returns an ordered list of subscriber protos with matching IDs.
+func (l *SubProtoStore) GetByIDs(network string, sids []string) ([]*lte_protos.SubscriberData, error) {
```

Comment:
Consider adding a guard that "if len(sids) > max page size then return err" -- and document that in the docstring

Reply:
The storage functions don't have access to the max page size config (for the `subscriberdb` servicer), but maybe this guard can be added in the `SyncSubscribers` handler of the servicer? @hcgatewood

Reply:
That's fine, just make a note of it here and then can enforce it in SyncSubsc